### ISSUE TITLE
linkerd upgrade --ha fails with "failed to read CA: not PEM-encoded"

### ISSUE THREAD

**Issue by siggy** [linkerd/linkerd2#5385]
### What is the issue?

`linkerd upgrade --ha` fails on `edge-20.12.3`

### How can it be reproduced?

Command succeeds prior to this commit, fails with it: https://github.com/linkerd/linkerd2/commit/35612ae26823a9dfec51facc927f1b68051cd255

```bash
# works
$ git checkout 377b38f0bfb83421a3b74434aea43188810a0d01
$ go run cli/main.go upgrade --ha
---
###
### Linkerd Namespace
...

# breaks on the next commit
$ git checkout 35612ae26823a9dfec51facc927f1b68051cd255
$ go run cli/main.go upgrade --ha
Error: failed to validate issuer credentials: failed to read CA: not PEM-encoded
Usage:
...
```

#### `linkerd check` output

```bash
$ linkerd check --context aks-dive-staging
kubernetes-api
--------------
√ can initialize the client
...
√ issuer cert is within its validity period
‼ issuer cert is valid for at least 60 days
    issuer certificate will expire on 2021-01-27T08:02:19Z
    see https://linkerd.io/checks/#l5d-identity-issuer-cert-not-expiring-soon for hints
√ issuer cert is issued by the trust anchor
...
control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 20.12.1 but the latest edge version is 20.12.3
    see https://linkerd.io/checks/#l5d-version-control for hints
‼ control plane and cli versions match
    control plane running edge-20.12.1 but cli running edge-20.12.3
    see https://linkerd.io/checks/#l5d-version-control for hints
...
Status check results are √
```

### Environment

- Kubernetes Version: v1.18.2
- Cluster Environment: AKS
- Host OS: Linux
- Linkerd version: edge-20.12.3

### Additional context

We use cert-manager.

**Comment by alpeb** [comment]
@siggy from my testing, I do get the same error with code bases prior to `35612ae` :man_shrugging: 

The error does make sense to me though... I assume you installed via Helm in order to set up cert-manager, via `--set identity.issuer.scheme=kubernetes.io/tls`? After 2.9 it's no longer possible to install via Helm and then upgrade through the CLI. This was done in order to simplify the upgrade logic. When upgrading via Helm, Helm's will use its own mechanism (the `sh.helm.release.v1.linkerd.v1` secret) to retrieve the template values and reuse them. The CLI uses a separate mechanism (the `linkerd-config-overrides` secret) so when you attempt a CLI upgrade on a Helm install, the `identity.issuer.scheme` config can't be retrieved and reading the issuer cert fails.

It might be worth detecting this type of cases to issue a better error message for sure.

**Comment by zaharidichev** [comment]
@alpeb @siggy We have not used Helm at all. When transitioning to cert manager, what I did is uninstall Linkerd and then install it through the CLI again with the external issuer option. So is there a reason then for this not to work?

**Comment by acobaugh** [comment]
Seeing the same error when upgrading with the CLI from 2.9.0 to 2.9.1:

```
% linkerd-2.9.1 upgrade | kubectl apply --prune -l linkerd.io/control-plane-ns=linkerd -f -                                                                                                                         
Error: failed to validate issuer credentials: failed to read CA: not PEM-encoded
```

This linkerd installation was installed may minor versions ago, and we've only ever used the CLI. We installed it thusly, after pre-creating the CA and issuer certs and keys:
```
linkerd install \
  --ha \
  --identity-trust-anchors-file certs/linkerd/ca.crt \
  --identity-issuer-certificate-file certs/linkerd/issuer.crt \
  --identity-issuer-key-file certs/linkerd/issuer.key \
  | kubectl apply -f -
```

I saw #5263, but we don't have a `secret/linkerd-config-overrides`, so maybe unrelated.

**Comment by alpeb** [comment]
@acobaugh I couldn't reproduce your issue by installing 2.9.0 using the command you pointed out and then upgrading to 2.9.1. Are you sure you didn't use `--identity-external-issuer` in your original install?  You can find out by looking into `kubectl get cm linkerd-config -oyaml` under `data.identity.issuer.scheme` (should be `kubernetes.io/tls` if you did use an external issuer).

**Comment by acobaugh** [comment]
My scheme is `linkerd.io/tls` in the configmap.

**Comment by alpeb** [comment]
We found there's indeed a bug when using the `--ha` in `linkerd upgrade`. Although your issue @acobaugh I still can't figure out. To begin with, when you upgraded to `2.9.0` the `linkerd-config-overrides` secret should have been created. How does `linkerd check` look like?

**Comment by acobaugh** [comment]
`linkerd check` all checks out. If this is a different bug, I can create another issue rather than hijack this one.


```
% linkerd-2.9.1 check
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist
linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 2.9.0 but the latest stable version is 2.9.1
    see https://linkerd.io/checks/#l5d-version-control for hints
‼ control plane and cli versions match
    control plane running stable-2.9.0 but cli running stable-2.9.1
    see https://linkerd.io/checks/#l5d-version-control for hints

linkerd-ha-checks
-----------------
√ multiple replicas of control plane pods
linkerd-prometheus
------------------
√ prometheus add-on service account exists
√ prometheus add-on config map exists
√ prometheus pod is running

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √
```

**Comment by alpeb** [comment]
@acobaugh thanks for the prompt reply. I could finally reproduce your case as long as I deleted the `linkerd-config-overrides` secret, and the issue is caused by the same bug I mentioned above. However I still can't figure how when you upgraded to `2.9.0` that secret wasn't created. As of that version, the issuer cert key is stored in there (and not in the `linkerd-config` ConfigMap). Without it, I'm afraid you require to install linkerd again from scratch to pass on those certs and key again (please uninstall beforehand using `linkerd uninstall | kubectl delete -f -` for a proper cleanup). Let me know how it goes.

**Comment by acobaugh** [comment]
We have 5 clusters (2 EKS, 3 on-prem) that were all updated to 2.9.0 that are missing that secret. CLI was used every time, and to the best of my knowledge the same process was used to install and update (in fact, two different people did the updates to 2.9.0). 

Just now I was able to reproduce it. Order of events was roughly
* linkerd-2.8.1 install
* linkerd 2.9.0 upgrade
* linkerd 2.9.1 upgrade

At no point did I see `secret/linkerd-config-overrides`. 

I did inadvertently go from 2.8.1 straight to 2.9.1, and that worked just fine.

For sake of completeness, this is the complete "bootstrap" script our system uses to create certs and install linkerd on a fresh cluster:

```
#!/bin/bash

cmd=step
[[ $(type -P "$cmd") ]] && echo "$cmd is in PATH"  ||
    { echo "$cmd is NOT in PATH" 1>&2; exit 1; }

mkdir -p certs/linkerd

step certificate create identity.linkerd.cluster.local certs/linkerd/ca.crt certs/linkerd/ca.key --profile root-ca --no-password --insecure
step certificate create identity.linkerd.cluster.local certs/linkerd/issuer.crt certs/linkerd/issuer.key --ca certs/linkerd/ca.crt --ca-key certs/linkerd/ca.key --profile intermediate-ca --not-after 87600h --no-password --insecure

linkerd install \
  --ha \
  --identity-trust-anchors-file certs/linkerd/ca.crt \
  --identity-issuer-certificate-file certs/linkerd/issuer.crt \
  --identity-issuer-key-file certs/linkerd/issuer.key \
  | kubectl apply -f
```

**Comment by alpeb** [comment]
@acobaugh Ok I see the problem was the `--prune` option used in `linkerd-2.9.1 upgrade | kubectl apply --prune -l linkerd.io/control-plane-ns=linkerd -f - `. Turns out the new secret doesn't have that label so it's not surviving that pruning... I would have thought kubectl only pruned resources _already_ living in the cluster, but it appears it also applies to resources being created :man_shrugging: 

I'll get that fixed. As a temporary workaround you can pipe `linkerd upgrade`'s output to a file, and manually add the `linkerd.io/control-plane-ns=linkerd` label to the `linkerd-config-overrides` secret, and then send that to `kubectl apply --prune ...`.

Thanks for the detailed feedback!

**Comment by adleong** [comment]
This fix should be backported to 2.9.3
